In [2]:
%load_ext autoreload
%autoreload 2

from config import cfg
from huggingface_hub import login
login(token=cfg.hf_token)

Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


In [9]:
jd = """
JOB DESCRIPTION
Job: Fresher AI Engineer
Location: FPT Software Building, No. 12 Science Avenue, Quy Nhon Nam Ward, Gia
Lai Province
Onboard: 3/2026
UNIT INTRODUCTION
Quy Nhon AI Research and Application Center (QAI) is one of FPT Software’s most
strategic units in artificial intelligence research and application. Our vision is to become
the leading AI Creative Center in Vietnam and ASEAN by 2030.
We offer a professional, innovative workplace where young engineers can grow rapidly
and gain hands-on experience through global projects (Japan, US, Canada, etc.).
JOB DESCRIPTION
- Receive intensive training to build advanced knowledge in AI/Generative AI
- Support the development, experimentation, and optimization of AI/Generative AI
models/solution under the guidance of senior engineers or mentors.
- Assist in data processing, model training, evaluation, and improvement following
project workflows.
- Research and explore new AI/LLM technologies and contribute ideas for applying
them to real-world problems.
- Collaborate with researchers, data scientists, and software engineers to integrate AI
models into applications or production systems.
- Prepare basic technical documentation related to model design, experiments, and
results.
- Participate in building AI solution for customer or internal innovation projects.
JOB SPECIFICATION
Technical Knowledge & Skills
- Bachelor’s degree (or final-year student) in Computer Science, Software Engineering,
Mathematics–Informatics, or related fields.
- Fundamental knowledge of:
    o Machine Learning, Deep Learning
    o NLP, Computer Vision
- Basic familiarity with AI frameworks such as PyTorch or TensorFlow.
- Understanding of Large Language Models (LLMs) or Generative AI is a strong
advantage.
- Experience through academic projects, research, capstone or personal AI projects is
preferred.
Technical Tools & Platforms
- Proficiency in Python and common data/ML libraries (NumPy, Pandas, Scikit-learn).
- Familiarity with Git/GitHub.
- (Optional but beneficial) Basic knowledge of cloud AI platforms such as Azure AI,
Google Vertex AI, or AWS SageMaker.
Soft Skills
- Strong logical thinking, willingness to learn, and proactive attitude.
- Good teamwork and communication skills.
- Ability to read and understand English technical documents.
-BENEFITS
- Direct mentorship and training by senior AI experts at QAI.
- Opportunity to work on real-world AI projects with international clients.
- Access to modern AI research infrastructure and the latest Generative AI/LLM
technologies.
- Training Benefits: Receive a full-course training allowance up to 24,000,000 VND,
depending on your contribution and training performance.
After completing the training program:
- Join real projects with a competitive salary aligned with your skills and experience,
ranging from 12–18 million VND/month…
CONTACT
- Sent CV to Email: NgocNTH6@fpt.com
"""

In [10]:
system_prompt_gen_question = """You are a Technical Recruiter and Senior AI Engineer.
Your task based on the following raw Job Description (JD), generate exactly 3 interview questions and their corresponding sample answers.
Constraints output must strictly follow this JSON format: [{"level": "Easy", "question": "...", "sample_answer": "..."}, ...].
Must include 1 Easy, 1 Medium, and 1 Hard question.
Focus ONLY on the core technical keywords mentioned in the JD."""

## Llama

In [46]:
import os
from huggingface_hub import InferenceClient

client = InferenceClient(
    api_key=os.environ["HF_TOKEN"],
)

completion = client.chat.completions.create(
    model="meta-llama/Llama-3.2-1B-Instruct:novita",
    messages = [
    {"role": "system", "content": system_prompt_gen_question},
    {"role": "user", "content": jd}],
)
print(completion.choices[0].message)

ChatCompletionOutputMessage(role='assistant', content='Here are three interview questions with sample answers based on the provided Job Description.\n\n**Easy Question:**\n\n**Question:** Can you explain the key differences between supervised and unsupervised machine learning algorithms?\n\n**Sample Answer:**\n\n"Thank you for the question. Supervised machine learning algorithms, such as linear regression, decision trees, and random forests, are designed to learn from labeled data and produce a predicted output based on the input features. In contrast, unsupervised machine learning algorithms, like k-means clustering and hierarchical clustering, aim to discover patterns and groupings in unlabeled data without any prior knowledge or labels. \n\nFor example, in a text classification problem, supervised algorithms like logistic regression or SVM would learn from labeled data with labels such as \'positive\' or \'negative\', and produce a predicted probability of the text being positive. O

ChatCompletionOutputMessage(role='assistant', content='Here are three interview questions with sample answers based on the provided Job Description.\n\n**Easy Question:**\n\n**Question:** Can you explain the key differences between supervised and unsupervised machine learning algorithms?\n\n**Sample Answer:**\n\n"Thank you for the question. Supervised machine learning algorithms, such as linear regression, decision trees, and random forests, are designed to learn from labeled data and produce a predicted output based on the input features. In contrast, unsupervised machine learning algorithms, like k-means clustering and hierarchical clustering, aim to discover patterns and groupings in unlabeled data without any prior knowledge or labels. \n\nFor example, in a text classification problem, supervised algorithms like logistic regression or SVM would learn from labeled data with labels such as \'positive\' or \'negative\', and produce a predicted probability of the text being positive. On the other hand, unsupervised algorithms like k-means clustering would group similar text features together without any prior knowledge of their labels."\n\n**Medium Question:**\n\n**Question:** How would you optimize the performance of a Deep Neural Network (DNN) architecture using a technique called dropout regularization?\n\n**Sample Answer:**\n\n"Dropout regularization is a technique used to prevent overfitting in deep neural networks by randomly dropping out fully connected layers during training. During training, dropout randomly sets a fraction of the input features to zero during each mini-batch, thus preventing the network from relying too heavily on any single input feature.\n\nTo optimize the performance of a DNN architecture using dropout regularization, we would follow these steps:\n\n1. **Batch normalization**: Before applying dropout, we would apply batch normalization to the input features to normalize the input dimensions and reduce the internal covariate shift.\n2. **Dropout**: We would randomly set a fraction of the input features to zero during each mini-batch, which would be a fraction of the total number of input features.\n3. **Gradient clipping**: We would clip the gradients of the loss function to prevent exploding gradients during backpropagation.\n4. **Weight decay**: We would add a weight decay term to the loss function to prevent overfitting.\n\nBy applying these techniques, we would effectively reduce overfitting and improve the generalization performance of the DNN architecture."\n\n**Hard Question:**\n\n**Question:** Design a Generative Adversarial Network (GAN) model to generate high-quality synthetic images of a specific object, such as a car, using a Generative Adversarial Network architecture.\n\n**Sample Answer:**\n\n"Here\'s a high-level design for a GAN model to generate high-quality synthetic images of a car:\n\n**Architecture:**\n\n1. **Generator**: We would use a U-Net architecture as the generator, which consists of an encoder-decoder structure with skip connections to reduce spatial dimensions.\n2. **Discriminator**: We would use a U-Net architecture as the discriminator, which consists of an encoder-decoder structure with skip connections to reduce spatial dimensions.\n3. **Generator**: The generator takes an input image and produces a synthetic image as its output.\n4. **Discriminator**: The discriminator takes an input image and produces a probability that the input image is real.\n5. **Training**: We would train the generator using adversarial training, where the generator is trained to produce synthetic images that are indistinguishable from real images, while the discriminator is trained to distinguish between real and synthetic images.\n\n**Training Process:**\n\n1. We would initialize the generator and discriminator using a random initialization.\n2. We would train the generator using adversarial training, where the generator is trained to produce synthetic images that are indistinguishable from real images.\n3. We would train the discriminator using adversarial training, where the discriminator is trained to distinguish between real and synthetic images.\n4. We would alternate between training the generator and discriminator to update their parameters.\n\n**Evaluation:**\n\nWe would evaluate the performance of the GAN model using metrics such as Mean Squared Error (MSE) and Frechet Inception Distance (FID).\n\n**Hyperparameters:**\n\nWe would choose the following hyperparameters for the GAN model:\n\n* Generator architecture: U-Net with skip connections\n* Discriminator architecture: U-Net with skip connections\n* Generator learning rate: 0.001\n* Discriminator learning rate: 0.001\n* Batch size: 32\n* Number of iterations: 10000\n\nBy following these steps and using these hyperparameters, we would effectively design a GAN model to generate high-quality synthetic images of a car using a Generative Adversarial Network architecture."', reasoning=None, tool_call_id=None, tool_calls=None)


In [ ]:
# Meta-Llama-3-8B-Instruct
from huggingface_hub import InferenceClient

client = InferenceClient(
    provider="auto",
    api_key=cfg.hf_token
)

completion = client.chat.completions.create(
    model="meta-llama/Meta-Llama-3-8B-Instruct",
    messages = [
    {"role": "system", "content": system_prompt_gen_question},
    {"role": "user", "content": jd},
],
    max_tokens=512,
)

print(completion.choices[0].message["content"])


Here are 3 interview questions with their corresponding sample answers, focusing on the core technical keywords mentioned in the JD:

[{"level": "Easy", "question": "What is the fundamental difference between Machine Learning and Deep Learning?", "sample_answer": "Machine Learning is a subset of AI that enables systems to learn from data without being explicitly programmed. Deep Learning is a type of Machine Learning that uses neural networks with multiple layers to analyze and interpret data. This allows Deep Learning models to learn complex patterns and relationships in data, making it particularly effective for image and speech recognition tasks."}, 
{"level": "Medium", "question": "How would you implement a basic Natural Language Processing (NLP) pipeline using PyTorch, including text preprocessing, tokenization, and sentiment analysis?", "sample_answer": "To implement a basic NLP pipeline using PyTorch, I would start by importing the necessary libraries and loading the dataset. Th

In [33]:
import os
from huggingface_hub import InferenceClient

client = InferenceClient(
    api_key=os.environ["HF_TOKEN"],
)

completion = client.chat.completions.create(
    model="meta-llama/Llama-3.1-8B-Instruct:novita",
    messages = [
    {"role": "system", "content": system_prompt_gen_question},
    {"role": "user", "content": jd},
],
)

completion.choices[0].message

ChatCompletionOutputMessage(role='assistant', content='Here are three interview questions for the Fresher AI Engineer position, categorized by difficulty level:\n\n[{"level": "Easy", "question": "What is the difference between supervised and unsupervised learning in machine learning?", "sample_answer": "In supervised learning, the model is trained on labeled data to predict the output for new, unseen data. On the other hand, in unsupervised learning, the model is trained on unlabeled data to discover patterns or relationships within the data."},\n{"level": "Medium", "question": "How would you implement a basic Natural Language Processing (NLP) task using PyTorch, such as sentiment analysis?", "sample_answer": "To implement sentiment analysis using PyTorch, I would first prepare the dataset, then create a PyTorch model using a pre-trained transformer such as BERT or RoBERTa. I would then fine-tune the model on the dataset using a custom dataset class and train the model using the PyTorc

ChatCompletionOutputMessage(role='assistant', content='Here are three interview questions for the Fresher AI Engineer position, categorized by difficulty level:\n\n[{"level": "Easy", "question": "What is the difference between supervised and unsupervised learning in machine learning?", "sample_answer": "In supervised learning, the model is trained on labeled data to predict the output for new, unseen data. On the other hand, in unsupervised learning, the model is trained on unlabeled data to discover patterns or relationships within the data."},\n{"level": "Medium", "question": "How would you implement a basic Natural Language Processing (NLP) task using PyTorch, such as sentiment analysis?", "sample_answer": "To implement sentiment analysis using PyTorch, I would first prepare the dataset, then create a PyTorch model using a pre-trained transformer such as BERT or RoBERTa. I would then fine-tune the model on the dataset using a custom dataset class and train the model using the PyTorch Trainer API. Finally, I would evaluate the model\'s performance using a metrics such as accuracy or F1-score."},\n{"level": "Hard", "question": "Design and implement a Generative Adversarial Network (GAN) using TensorFlow to generate images from a given dataset. How would you optimize the model to improve its performance?", "sample_answer": "To implement a GAN using TensorFlow, I would first create a TensorFlow dataset from the given dataset, then define the generator and discriminator networks. I would then use the Adam optimizer and a binary cross-entropy loss function to train the model. To optimize the model, I would use techniques such as batch normalization, learning rate scheduling, and weight clipping to stabilize the training process and improve the model\'s performance."}]', reasoning=None, tool_call_id=None, tool_calls=None)

## Qwen

In [ ]:
# Qwen2.5-0.5B
from transformers import AutoModelForCausalLM, AutoTokenizer

model_name = "Qwen/Qwen2.5-0.5B-Instruct"

model = AutoModelForCausalLM.from_pretrained(
    model_name,
    torch_dtype="auto",
    device_map="auto"
)
tokenizer = AutoTokenizer.from_pretrained(model_name)

Loading weights:   0%|          | 0/290 [00:00<?, ?it/s]

In [13]:
messages = [
    {"role": "system", "content": system_prompt_gen_question},
    {"role": "user", "content": jd}
]
text = tokenizer.apply_chat_template(
    messages,
    tokenize=False,
    add_generation_prompt=True
)
model_inputs = tokenizer([text], return_tensors="pt").to(model.device)

generated_ids = model.generate(
    **model_inputs,
    max_new_tokens=512
)
generated_ids = [
    output_ids[len(input_ids):] for input_ids, output_ids in zip(model_inputs.input_ids, generated_ids)
]

response = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)[0]
print(response)

### Easy Question 1:
"Could you please share an example of a project you have worked on that utilized machine learning and deep learning techniques?"

**Sample Answer:**
Certainly! One of my recent projects involved developing a recommendation system for a social media platform. We used a combination of collaborative filtering algorithms and unsupervised learning techniques to suggest users who might be interested in content they haven't seen yet.

### Easy Question 2:
"What are some key components of a typical day in an AI engineering role at QAI?"

**Sample Answer:**
In my role, I would spend most days working closely with senior engineers and mentors to guide and mentor junior engineers. My primary responsibilities include:

1. **Data Processing**: Handling large datasets using Python libraries like NumPy and Pandas.
2. **Model Training**: Using popular frameworks like PyTorch or TensorFlow for training AI models.
3. **Evaluation**: Monitoring model performance during training and d

## mistral

In [31]:
import os
from openai import OpenAI

client = OpenAI(
    base_url="https://router.huggingface.co/v1",
    api_key=os.environ["HF_TOKEN"],
)

prompt = f"""System: {system_prompt_gen_question}
User: {jd}"""
completion = client.chat.completions.create(
    model="mistralai/Mistral-7B-Instruct-v0.2:featherless-ai",
    messages=[
        {
            "role": "user",
            "content": prompt
        }
    ],
    max_tokens= 512
)

print(completion.choices[0].message)

ChatCompletionMessage(content='Based on the provided Job Description, here are three interview questions and corresponding sample answers for an Fresher AI Engineer:\n\nQuestion 1 (Easy Level):\n"Can you explain what Deep Learning is, and give an example of a Deep Learning model?"\n\nSample Answer:\n"Deep Learning is a subfield of Machine Learning, where the model consists of multiple layers, allowing the network to automatically learn and extract features from data, without explicit programming.\n\nFor example, a well-known Deep Learning model is AlexNet, which was published in 2012 and won the ImageNet Large Scale Visual Recognition Challenge that year. AlexNet consists of eight layers and uses ReLU activation functions, pooling layers, and dropout rate to prevent overfitting. AlexNet achieved a top-5 error rate of 15.3% on the ImageNet dataset, which was a significant improvement from the previous state-of-the-art model, which achieved a top-5 error rate of 26.2%."\n\nQuestion 2 (Me

ChatCompletionMessage(content='Based on the provided Job Description, here are three interview questions and corresponding sample answers for an Fresher AI Engineer:\n\nQuestion 1 (Easy Level):\n"Can you explain what Deep Learning is, and give an example of a Deep Learning model?"\n\nSample Answer:\n"Deep Learning is a subfield of Machine Learning, where the model consists of multiple layers, allowing the network to automatically learn and extract features from data, without explicit programming.\n\nFor example, a well-known Deep Learning model is AlexNet, which was published in 2012 and won the ImageNet Large Scale Visual Recognition Challenge that year. AlexNet consists of eight layers and uses ReLU activation functions, pooling layers, and dropout rate to prevent overfitting. AlexNet achieved a top-5 error rate of 15.3% on the ImageNet dataset, which was a significant improvement from the previous state-of-the-art model, which achieved a top-5 error rate of 26.2%."\n\nQuestion 2 (Medium Level):\n"Suppose you are given a dataset containing images of various animals (e.g., lions, tigers, elephants, zebras, etc.). Describe how you would preprocess the data, choose appropriate feature extraction techniques, select a suitable deep learning model architecture, and train and validate the model on this dataset."\n\nSample Answer:\n"To preprocess the data, I would resize all the images to a fixed size (e.g., 224x224 pixels for the input size required by the popular deep learning model architectures like ResNet, VGG, and Inception). I would also perform data augmentation techniques like random horizontal flips, random vertical flips, and random rotations to increase the size of the dataset and improve the model\'s ability to learn and generalize to new data."\n\n"For feature extraction techniques, I would explore using pre-trained convolutional neural network (CNN) models like VGG16, ResNet50, Inception V3, or Google\'s MobileNet to extract high-level features from the input images. These features would capture the spatial and contextual relationships present in the images and would provide a more compact and abstract representation of the data that could be used as input to the deep learning model."\n\n"For the deep learning', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None)